lib/collection/src/shards/replica_set/mod.rs

```diff
@@ -369,10 +373,16 @@ impl ShardReplicaSet {
     /// Wait for a replica set state condition to be true.
     ///
     /// Uses a blocking thread internally.
+    ///
+    /// # Cancel safety
+    ///
+    /// This method is cancel safe.
     async fn wait_for<F>(&self, check: F, timeout: Duration) -> CollectionResult<()>
     where
         F: Fn(&ReplicaSetState) -> bool + Send + 'static,
     {
+        // TODO: Propagate cancellation into `spawn_blocking` task!?
```

Comment:
Probably a good idea. It currently relies on the timeout to eventually drop the thread.

Moving the cancellation flag in and checking on each loop is probably already good enough.

Reply:
It's just rather non-trivial to do (because we have a sync task blocked on a cond-var), so I left it as `TODO` for now.